src/briefcase/integrations/subprocess.py

```diff
@@ -235,26 +294,46 @@ def stream_output(self, label, popen_process):
         :param label: A description of the content being streamed; used for
             to provide context in logging messages.
         :param popen_process: a running Popen process with output to print
+        :param stop_func: a Callable that returns True when output streaming
+            should stop and the popen_process should be terminated.
         """
+        output_streamer = threading.Thread(
+            name=f"{label} output streamer",
+            target=self._stream_output_thread,
+            args=(popen_process,),
+        )
         try:
-            while True:
-                # readline should always return at least a newline (ie \n)
-                # UNLESS the underlying process is exiting/gone; then "" is returned
-                output_line = ensure_str(popen_process.stdout.readline())
-                if output_line:
-                    self.command.logger.info(output_line)
-                elif output_line == "":
-                    # a return code will be available once the process returns one to the OS.
-                    # by definition, that should mean the process has exited.
-                    return_code = popen_process.poll()
-                    # only return once all output has been read and the process has exited.
-                    if return_code is not None:
-                        self._log_return_code(return_code)
-                        return
-
+            output_streamer.start()
+            if stop_func:
+                while output_streamer.is_alive() or not stop_func():
+                    time.sleep(0.5)
+            else:
+                output_streamer.join()
         except KeyboardInterrupt:
+            pass  # allow CTRL+C to gracefully stop streaming
+        finally:
             self.cleanup(label, popen_process)
```

Comment:
I moved `cleanup()` in to a `finally` block; my changes had originally introduced multiple exit points for a deceptively simple loop where `cleanup()` would need to be invoked for some exits but not others....although, it was definitely not clear when/why to call `cleanup()` unless you're pretty familiar with all the interactions.

The `cleanup()` function is safe to call multiple times and largely implements short circuits for already dead processes. I talk about this more with the changes to `cleanup` itself below.